core/src/main/java/com/linecorp/armeria/internal/client/dns/DefaultDnsResolver.java

```diff
@@ -116,50 +104,29 @@ private CompletableFuture<List<DnsRecord>> resolveOne(DnsQuestionContext ctx, Dn
     @VisibleForTesting
     CompletableFuture<List<DnsRecord>> resolveAll(DnsQuestionContext ctx, List<? extends DnsQuestion> questions,
                                                   String logPrefix) {
-
-        final List<CompletableFuture<List<DnsRecord>>> results =
-                questions.stream()
-                         .map(question -> delegate.resolve(ctx, question))
-                         .collect(toImmutableList());
-
-        final CompletableFuture<List<DnsRecord>> future =
-                CompletableFuture.allOf(results.toArray(EMPTY_FUTURES)).handle((unused0, unused1) -> {
-                    final List<DnsRecord> records = new ArrayList<>();
-                    List<Throwable> causes = null;
-                    for (CompletableFuture<List<DnsRecord>> result : results) {
-                        try {
-                            records.addAll(result.get());
-                        } catch (Throwable ex) {
-                            if (records.isEmpty()) {
-                                if (causes == null) {
-                                    causes = new ArrayList<>();
-                                }
-                                causes.add(Exceptions.peel(ex));
-                            }
-                        }
-                    }
-
-                    if (!records.isEmpty()) {
-                        if (records.size() > 1) {
-                            records.sort(Comparator.comparing(DnsRecord::type, preferredOrder));
-                        }
-                        return Collections.unmodifiableList(records);
-                    }
-
-                    final Throwable cause;
-                    if (causes == null) {
-                        cause = new UnknownHostException("Failed to resolve: " + questions + " (empty result)");
-                    } else {
-                        cause = new UnknownHostException("Failed to resolve: " + questions);
-                        for (Throwable c : causes) {
-                            cause.addSuppressed(c);
-                        }
-                    }
-                    return Exceptions.throwUnsafely(cause);
-                });
+        final CompletableFuture<List<DnsRecord>> future = new CompletableFuture<>();
+        final Object[] results = new Object[questions.size()];
+        for (int i = 0; i < questions.size(); i++) {
+            final int order = i;
+            delegate.resolve(ctx, questions.get(i)).handle((records, cause) -> {
+                assert executor.inEventLoop();
```

Comment:
While reviewing #4748, i'm wondering is it safe to assume this?
If `delegate` is a `CachingDnsResolver`, I guess this assertion might not be true

Reply:
Your analysis is correct.
I assumed an event loop thread invokes `DefaultDnsResolver` which turned out to be false.

Reply:
I actually meant that `CachingDnsResolver` is a cache which contains `CompletableFuture`s.
Depending on which `eventloop` completes the `CompletableFuture`, the original `eventloop` doesn't necessarily handle the callback

Reply:
The `CompletableFuture`s from `CachingDnsResolver` are all completed by the same `eventLoop` so isn't it okay?

Reply:
A `CachingDnsResolver` is executed by a `DefaultDnsResolver` that is executed by an event loop. Therefore, one event loop accesses `CachingDnsResolver`.